File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/javascript/reverse_string.js
function reverseString(str) {
  // type your code here
  const strArray = str.split("")
  let strHash = {}
  strArray.forEach((value, index) => {
    strHash[index] = value
  })
  reverseStrArray = []
  for (i = strArray.length - 1; i >= 0; i--) {
    reverseStrArray.push(strHash[i])
  }

  return (!!strArray.length > 0 ? reverseStrArray.join("") : str)
  
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'ih'");
  console.log("=>", reverseString("hi"));

  console.log("");

  console.log("Expecting: 'ybabtac'");
  console.log("=>", reverseString("catbaby"));
}

module.exports = reverseString;

// Please add your pseudocode to this file
/*
  -take in a string
  -split string into an array with an index for each character
  -map this array to a hash, with the indexes being the keys
  -map this hash back to an array, with the keys being the indexes in reverse order
*/

// And a written explanation of your solution
/*
  The initial string will have to be iterated over in order to be broken into an array, this array will be
  iterated over again in order to be turned into a hash, this hash will then be iterated over again in order
  to be transformed back into an array.  Once this final array is merged back into a string this string will be
  returned.  I believe this solution will always involve this many iterations no matter how long the string is.
*/

File: 01-week-1--starter-algorithms/00-day-1--reverse-a-string/ruby/reverse_string.rb
def reverse_string(str)
  # type your code in here
  str_array = str.split("")
  reversed_str = ""
  str_array.each do |char|
    reversed_str = char + reversed_str
  end
  reversed_str
end


if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'ih'"
  puts "=>", reverse_string('hi')

  puts 

  puts "Expecting: 'ybabtac'"
  puts "=>", reverse_string('catbaby')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
=begin
  -take in a string, convert it to an array
  -declare a variable as an empty string
  -iterate over the created array, adding each element to the start of the empty string variable and redefining that
   variable
  -return the variable
=end
# And a written explanation of your solution
=begin
  creating an array of the string seems like the easiest way to iterate over the characters of the string.
  at first I thought this array would then have to be converted to a hash in order to take advantage of hash
  lookup times, and then initiate another "reversed_array" and use a counter to iterate across the keys of the hash
  using a while loop and assigning these keys from greatest to least to the reversed_array.  Then return the joined
  reversed array.
=end